library/tlconfig.py

```diff
@@ -31,41 +32,52 @@
 TLCONFIG = "/opt/thinlinc/bin/tl-config"
 
 def getvalue(param):
-    tlproc = subprocess.Popen([TLCONFIG, "%s" % param], stdout=subprocess.PIPE)
-    stdout, _ = tlproc.communicate()
-    return stdout.rstrip()
+    tlproc = subprocess.run([TLCONFIG, "%s" % param], stdout=subprocess.PIPE, universal_newlines=True)
+    return tlproc.stdout.strip("\n")
 
-def setvalue(param, value):
-    tlproc = subprocess.Popen([TLCONFIG, "%s=%s" % (param, value)])
-    return tlproc.wait()
 
-def main():
-    changed = False
+def setvalue(param, value, check_mode):
+    if not check_mode:
+        tlproc = subprocess.run([TLCONFIG, "%s=%s" % (param, value)])
+        return tlproc.returncode
+    else:
+        return 0
 
+def main():
     module = AnsibleModule(argument_spec=dict(param=dict(required=True, type='str'),
                                               value=dict(required=True)),
                            supports_check_mode=True)
 
-    if module.check_mode:
-        module.exit_json(changed=module.params['value'] != getvalue(module.params['param']))
+    result = {
+        "changed": False,
+        "param": module.params['param'],
+        "value": module.params['value']
+    }
 
     if len(module.params['param']) == 0:
         module.fail_json(msg="Param can't be empty")
 
     if len(module.params['value']) == 0:
         module.fail_json(msg="Value can't be empty")
 
-    if getvalue(module.params['param']) != module.params['value']:
-        ret = setvalue(module.params['param'], module.params['value'])
+    current = getvalue(module.params['param'])
+    new = module.params['value']
 
-        if ret is not 0:
+    if current != new:
+        ret = setvalue(module.params['param'], module.params['value'], module.check_mode)
+
+        if ret != 0:
             msg = "Failed setting %s: tl-config returned %d" % (module.params['param'], ret)
             module.fail_json(msg=msg)
-        changed = True
+        result['changed'] = True
+
+        if module._diff:
+            result['diff'] = {
+                "before": yaml.safe_dump(current),
+                "after": yaml.safe_dump(new)
```

Comment:
This looks very odd. Are you sure it's supposed to be a yaml structure inside a json structure? That seems very redundant.

Reply:
I'm not too sure, I used an example module for making the diff output and it used ```yaml.safe_dump``` to format the output. However, I have taken it out and it still seems to work just fine.

(sorry, I'm not a :snake: wizard ;-) )